File: sdks/csharp/src/Nanoservice.Core/Types/Context.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Nanoservice.Core.Types;

/// <summary>
/// Context represents the workflow execution context passed between nodes.
/// </summary>
public class Context
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("workflow_name")]
    public string WorkflowName { get; set; } = string.Empty;

    [JsonPropertyName("workflow_path")]
    public string WorkflowPath { get; set; } = string.Empty;

    [JsonPropertyName("request")]
    public Request Request { get; set; } = new();

    [JsonPropertyName("response")]
    public Response Response { get; set; } = new();

    [JsonPropertyName("vars")]
    public Dictionary<string, object?> Vars { get; set; } = new();

    [JsonPropertyName("env")]
    public Dictionary<string, string> Env { get; set; } = new();

    /// <summary>
    /// Store a variable in context for downstream nodes.
    /// </summary>
    public void SetVar(string key, object? value)
    {
        Vars[key] = value;
    }

    /// <summary>
    /// Retrieve a variable from context.
    /// </summary>
    public object? GetVar(string key)
    {
        return Vars.TryGetValue(key, out var value) ? value : null;
    }

    /// <summary>
    /// Retrieve a variable from context as a string.
    /// </summary>
    public string? GetVarString(string key)
    {
        var value = GetVar(key);
        return value?.ToString();
    }
}

File: sdks/php/src/Types/Context.php
<?php

declare(strict_types=1);

namespace Blok\Blok\Types;

/**
 * Context represents the workflow execution context passed between nodes.
 */
final class Context
{
    public function __construct(
        public string $id = '',
        public string $workflowName = '',
        public string $workflowPath = '',
        public Request $request = new Request(),
        public Response $response = new Response(),
        /** @var array<string, mixed> */
        public array $vars = [],
        /** @var array<string, string> */
        public array $env = [],
    ) {}

    /**
     * Store a variable in context for downstream nodes.
     */
    public function setVar(string $key, mixed $value): void
    {
        $this->vars[$key] = $value;
    }

    /**
     * Retrieve a variable from context.
     */
    public function getVar(string $key): mixed
    {
        return $this->vars[$key] ?? null;
    }

    /**
     * Retrieve a string variable from context.
     */
    public function getVarStr(string $key): ?string
    {
        $val = $this->vars[$key] ?? null;
        return is_string($val) ? $val : null;
    }

    /**
     * Create a Context from an associative array.
     */
    public static function fromArray(array $data): self
    {
        return new self(
            id: $data['id'] ?? '',
            workflowName: $data['workflow_name'] ?? '',
            workflowPath: $data['workflow_path'] ?? '',
            request: isset($data['request']) ? Request::fromArray($data['request']) : new Request(),
            response: isset($data['response']) ? Response::fromArray($data['response']) : new Response(),
            vars: $data['vars'] ?? [],
            env: $data['env'] ?? [],
        );
    }

    /**
     * Convert to an associative array for JSON serialization.
     */
    public function toArray(): array
    {
        return [
            'id' => $this->id,
            'workflow_name' => $this->workflowName,
            'workflow_path' => $this->workflowPath,
            'request' => $this->request->toArray(),
            'response' => $this->response->toArray(),
            'vars' => $this->vars,
            'env' => $this->env,
        ];
    }
}
